src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AQNBuilder.Mono.cs

```diff
@@ -0,0 +1,483 @@
+
+				m_parseState = ParseState.START;
+
+				if (m_bHasAssemblySpec)
+				{
+					if (m_bUseAngleBracketsForGenerics)
+						Append('>');
+					else
+						Append(']');
+				}
+
+				PopOpenGenericArgument();
+			}
+			public void AddName(string szName)
+			{
+				if (szName == null)
+					Fail ();
+
+				if (!CheckParseState(ParseState.START | ParseState.NAME))
+					Fail ();
+
+				m_parseState = ParseState.NAME;
+
+				if (m_bNestedName)
+					Append ('+');
+
+				m_bNestedName = true;
+
+				EscapeName (szName);
+			}
+
+			public void AddPointer()
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR))
+					Fail ();
+
+				m_parseState = ParseState.PTRARR;
+
+				Append ('*');
+			}
+
+			public void AddByRef()
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR))
+					Fail();
+
+				m_parseState = ParseState.BYREF;
+
+				Append('&');
+			}
+
+			public void AddSzArray()
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR))
+					Fail();
+
+				m_parseState = ParseState.PTRARR;
+
+				Append("[]");
+			}
+			public void AddArray(int rank)
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR))
+					Fail();
+
+				m_parseState = ParseState.PTRARR;
+
+				if (rank <= 0)
+					throw new ArgumentOutOfRangeException ();
+
+				if (rank == 1)
+					Append("[*]");
+				else if (rank > 64)
+				{
+					// Only taken in an error path, runtime will not load arrays of more than 32 dimentions
+					Append ($"[{rank}]");
+				}
+				else
+				{
+					var strDim = new StringBuilder (rank + 3);
+
+					strDim.Append ('[');
+					for(int i = 1; i < rank; i++) strDim.Append (',');
+					strDim.Append (']');
+					Append(strDim.ToString ());
+				}
+			}
+			public void AddAssemblySpec(string assemblySpec)
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR | ParseState.BYREF))
+					Fail();
+
+				m_parseState = ParseState.ASSEMSPEC;
+
+				if (assemblySpec != null && !assemblySpec.Equals(String.Empty))
+				{
+					Append(", ");
+
+					if (m_instNesting > 0)
+					{
+						EscapeEmbeddedAssemblyName(assemblySpec);
+					}
+					else
+					{
+						EscapeAssemblyName(assemblySpec);
+					}
+
+					m_bHasAssemblySpec = true;
+				}
+
+			}
+			public void ToString(out string retString)
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR | ParseState.BYREF | ParseState.ASSEMSPEC))
+					Fail();
+
+				if (m_instNesting == 0)
+					Fail();
+				retString = m_pStr.ToString ();
+			}
+
+			public void Clear()
+			{
+				if (m_pStr != null)
+				{
+					m_pStr.Clear ();
+				}
+				m_bNestedName = false;
+				m_instNesting = 0;
+				m_bFirstInstArg = false;
+				m_parseState = ParseState.START;
+				m_bHasAssemblySpec = false;
+				m_bUseAngleBracketsForGenerics = false;
+				// m_stack.Clear (); looks like Mono's Stack type is missing Clear()
```

Comment:
The System.Collections.Stack type is implemented in / exposed from System.Collections.NonGeneric.  Which Stack is this?
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Collections.NonGeneric/src/System/Collections/Stack.cs